tes-adapter/src/main/java/com/epam/pipeline/tesadapter/service/TaskMapper.java

```diff
@@ -65,8 +80,54 @@ public PipelineStart mapToPipelineStart(TesTask tesTask) {
     }
 
     private TesExecutor getExecutorFromTesExecutorsList(List<TesExecutor> tesExecutors) {
-        Assert.notEmpty(tesExecutors, messageHelper.getMessage(MessageConstants.ERROR_PARAMETER_NULL_OR_EMPTY,
-                "executors"));
+        Assert.isTrue(tesExecutors.size() == ONLY_ONE, messageHelper.getMessage(
+                MessageConstants.ERROR_PARAMETER_INCOMPATIBLE_CONTENT, EXECUTORS, tesExecutors));
         return tesExecutors.get(FIRST);
     }
+
+    private String getProperInstanceType(TesTask tesTask, Tool pipelineTool) {
+        Double ramGb = tesTask.getResources().getRamGb();
+        Long cpuCores = tesTask.getResources().getCpuCores();
+        Long toolId = pipelineTool.getId();
+        Long regionId = getProperRegionIdInCloudRegionsByTesZone(tesTask.getResources().getZones());
+        Boolean spot = tesTask.getResources().getPreemptible();
+
+        AllowedInstanceAndPriceTypes allowedInstanceAndPriceTypes = cloudPipelineAPIClient
+                .loadAllowedInstanceAndPriceTypes(toolId, regionId, spot);
+        Assert.isTrue(allowedInstanceAndPriceTypes.getAllowedInstanceTypes() != null, messageHelper.getMessage(
+                MessageConstants.ERROR_PARAMETER_NULL_OR_EMPTY, allowedInstanceAndPriceTypes));
+        return evaluateMostProperInstanceType(allowedInstanceAndPriceTypes, ramGb, cpuCores);
+    }
+
+    private Tool loadToolByTesImage(String image) {
+        return cloudPipelineAPIClient.loadTool(image);
+    }
+
+    private Long getProperRegionIdInCloudRegionsByTesZone(List<String> zones) {
+        Assert.isTrue(zones.size() == ONLY_ONE, messageHelper.getMessage(
+                MessageConstants.ERROR_PARAMETER_INCOMPATIBLE_CONTENT, ZONES, zones));
+        return cloudPipelineAPIClient.loadAllRegions().stream().filter(
+                region -> region.getName().equalsIgnoreCase(zones.get(FIRST)))
+                .collect(Collectors.toList()).get(FIRST).getId();
+    }
+
+    private String evaluateMostProperInstanceType(AllowedInstanceAndPriceTypes allowedInstanceAndPriceTypes,
+                                                  Double ramGb, Long cpuCores) {
+        return allowedInstanceAndPriceTypes.getAllowedInstanceTypes().stream()
+                .collect(Collectors.toMap(InstanceType::getName, instancaType ->
+                        calculateInstanceCoef(instancaType, ramGb, cpuCores))).entrySet().stream()
+                .min(Comparator.comparing(Map.Entry::getValue)).orElseThrow(NullPointerException::new).getKey();
+    }
+
+    private Double calculateInstanceCoef(InstanceType instanceType, Double ramGb, Long cpuCores) {
+        return Math.abs((instanceType.getMemory() * parseInstanceMemoryUnit(instanceType.getMemoryUnit())
+                / ramGb + instanceType.getVCPU() / cpuCores) / 2 - 1);
```

Comment:
to calculate this coefficient more accurate you need to cast VCPU and CpuCores to double, because if VCPU = 4 and cpuCores = 5,  instanceType.getVCPU() / cpuCores will be 0 in case of integer and long values

Reply:
fixed, (from point of clean-code) is it significant difference between to cast like "(double) value", or use "longValue.doubleValue()" for Long values?

Reply:
use longValue.doubleValue()